### PR TITLE
US1468 Design doc for replication protocol

### PR DESCRIPTION
**Pull request by jkryl**
Add new test cases mentioned in the design doc and fix the code for
test cases which did not pass.

### CODE REVIEW

**Comment by vishnuitta** [comment]
@jkryl .. listing testcases would be helpful for review. Can you please add that list?

**Comment by jkryl** [comment]
Hi Vitta, this is the list of test cases from the doc:

Management connection:
Handshake request with a zvol name which does not exist.
Handshake request with a zvol which exists (success).
Request with unknown opcode.
Request with unsupported version number.
Test connect retry logic.

Data connection:
Read a block which has not been previously written (without ionum)
Write a block with ionum.
Read a block and verify its ionum.
Read a block which consists of multiple data chunks with different ionums.
Issue a sync request.
A request with unknown opcode.
Read request with offset greater than zvol size.
Read request with offset + length greater than zvol size.
Write request with offset greater than zvol size.
Write request with offset + length greater than zvol size.
Read request on an online zvol with rebuild flag set.

Current gtests cover all cases mentioned above.

File: tests/cbtest/gtest/test_zrepl_prot.cc [hunk 1]
```diff
@@ -699,6 +734,269 @@ TEST_F(ZreplDataTest, ReadBlockWithoutMeta) {
...
+	ASSERT_ERRNO("read", rc >= 0);
+	ASSERT_EQ(rc, sizeof (hdr_in));
+	EXPECT_EQ(hdr_in.opcode, ZVOL_OPCODE_SYNC);
+	EXPECT_EQ(hdr_in.status, ZVOL_OP_STATUS_OK);
+	EXPECT_EQ(hdr_in.io_seq, m_ioseq);
+	EXPECT_EQ(hdr_in.offset, 0);
+	EXPECT_EQ(hdr_in.len, 0);
+}
+
+TEST_F(ZreplDataTest, UnknownOpcode) {
```
Comment (vishnuitta) [inline tests/cbtest/gtest/test_zrepl_prot.cc:777]: this seems to be already covered

**Review by vishnuitta — APPROVED** [review]
looks good. thanks

File: tests/cbtest/gtest/test_zrepl_prot.cc (hunk 1 above)
Comment (jkryl) [reply tests/cbtest/gtest/test_zrepl_prot.cc:777]: it's covered for Zrepl mgmt connection, but this one is for data connection.

Comment (vishnuitta) [reply tests/cbtest/gtest/test_zrepl_prot.cc:777]: sounds good.. waiting for travis to pass to do the merge

**Pull request merged**